app/src/main/java/ca/dal/cs/athletemonitor/athletemonitor/AccountManager.java

```diff
@@ -0,0 +1,226 @@
+                if (dataSnapshot.exists()) {
+                    User userLoggingIn = dataSnapshot.getValue(User.class);
+
+                    //compare passwords, if they match add the user to the logged in users list
+                    //and return a successful login attempt, otherwise, report a failed login
+                    if (userLoggingIn.getPassword().equals(user.getPassword())) {
+                        AccountManager.setUserLoginState(user.getUsername(), true);
+
+                        listener.onResult(true);
+                    } else {
+                        listener.onResult(false);
+                    }
+                } else {
+                    listener.onResult(false);
+                }
+            }
+
+            @Override
+            public void onCancelled(DatabaseError databaseError) {
+            }
+        });
+    }
+
+    /**
+     * Sets whether a user is marked as being online or not.  Users are tagged as online will
+     * have a node in the online_users branch of the database
+     *
+     * @param username Name of user to tag online/offline
+     * @param online Online state of user.  True if online, False if offline
+     */
+    public static void setUserLoginState(String username, boolean online) {
+        //retrieve a reference to the online users node
+        FirebaseDatabase database = FirebaseDatabase.getInstance();
+        DatabaseReference usersReference = database.getReference("online_users/");
+
+        //if the user is to be tagged as online, write an entry in the online users branch
+        //otherwise, delete any existing entry
+        if (online) {
+            usersReference.child(username).setValue(true);
+        } else {
+            usersReference.child(username).removeValue();
+        }
+    }
+
+    /**
+     * Determines whether or not the specified user exists in the database
+     *
+     * @param username Name of the user to look up
+     *
+     */
+    public static void userExists(final String username, final UserExistsListener listener) {
+        //retrieve a reference to the users node
+        FirebaseDatabase database = FirebaseDatabase.getInstance();
+        DatabaseReference usersReference = database.getReference("users/");
+
+        //attach a listener for data changes of the users reference.  this will occur when
+        //the reference is populated
+        usersReference.addListenerForSingleValueEvent(new ValueEventListener() {
+            @Override
+            public void onDataChange(DataSnapshot dataSnapshot) {
+                //if the username is a child of the users node then fire the user exists event
+                //indicating true, otherwise, fire event with false outcome
+                if (dataSnapshot.hasChild(username)) {
+                    if (listener != null) listener.onUserExists(true);
+                }
+                else {
+                    if (listener != null) listener.onUserExists(false);
+                }
```

Comment:
Lines 146-151 should be
` if(listener != null) {`
`listener.onUserExists(dataSnapshot.hasChild(username));`
`}`

However, why are you worried about listener being null here? You do not check this in all other places that you use ValueEventListener. At a glance, I don't think you need to check if it's not null, because if it's null then something is wrong and you don't want to silence that.

Whichever is the case, please either drop the `listener != null` checks here or add them everywhere else (and explain why).

Reply:
The null check is necessary since it is possible to call the methods without registering a callback.  This is a potential bug in the other callback code; will update them.

Reply:
Okay cool, my thought is just that right now, if the method is called without registering, things will go wrong and you won't know why. I'll leave it up to you, but in my opinion there are two options:
(1) Remove all checks of `if(listener != null) ...` so that a NullPointerException happens and the issue is exposed, or,
(2) To be really safe, it'd actually be best to throw an exception,
`if(listener == null) throw new IllegalStateException("Null callback listener");`

Reply:
@nickburris  Ack. I've adjusted a couple of them with annotation and non-null checks where it makes absolute sense like in the userExists method.

I believe not registering a callback might be a legitimate use case in some situations.  As an example, in the authenticate method, a user is marked as being online, so there might be cases where getting the actual result at the time is unnecessary thus causing unnecessary code bloat.   I would like to leave the rest as is for now, but take another look at it in the future.